src/frontend/org/voltdb/calciteadapter/rules/physical/VoltDBCalcAggregateMergeRule.java

```diff
@@ -0,0 +1,84 @@
+/* This file is part of VoltDB.
+ * Copyright (C) 2008-2018 VoltDB Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.voltdb.calciteadapter.rules.physical;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rex.RexLocalRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexProgram;
+import org.apache.calcite.rex.RexProgramBuilder;
+import org.voltdb.calciteadapter.rel.physical.VoltDBAggregate;
+import org.voltdb.calciteadapter.rel.physical.VoltDBCalc;
+
+/**
+ * This simply merges/inlines Calc's condition into an Aggregate node.
+ * In the VoltDB land this condition represents aggregate's HAVING clause
+ *
+ */
+public class VoltDBCalcAggregateMergeRule extends RelOptRule {
+
+    public static final VoltDBCalcAggregateMergeRule INSTANCE = new VoltDBCalcAggregateMergeRule();
+
+    private VoltDBCalcAggregateMergeRule() {
+        super(operand(VoltDBCalc.class,
+                operand(VoltDBAggregate.class, any())));
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        VoltDBCalc calc= call.rel(0);
+        // We only can merge a Calc's condition
+        return calc.getProgram().getCondition() != null;
```

Comment:
Calc can also have a projection component too?  If there is a projection component then I think it would not match, the calc node would need to be converted to a projection?

Reply:
If a Calc above an aggregate has a projection and condition, this rule will match because calc.getProgram().getCondition() won't be null. Why do you think it won't match?

The condition expression gets inlined with the VoltDBAggregate rel to be converted into a VoltDB aggregate's post predicate during the toPlanNode() call. A new VoltDBCalc rel containing only projection is created to replace the original VOltDBCalc. The new Calc will be converted to a ProjectionPlanNode during the same toPlanNode(). Its condition will be NULL so the toPlanNode should succeed.
Does it make sense?